src/ctia/entity/event/obj_to_event.clj

```diff
@@ -41,27 +41,146 @@
             :action "added"
             :change {}})
          (remove #(contains? diff-before %)
-                 (keys diff-after))))))
+                 (keys diff-after)))))
 
+(defn truncate
+  "Truncate v, a possibly-nested value, by inserting placeholder after given
+  max-depth and max-count."
+  ([v placeholder max-count max-depth] (truncate v placeholder max-count max-depth 0))
+  ([v placeholder max-count max-depth depth]
+   (let [tru #(truncate % placeholder max-count max-depth (inc depth))
+         dots placeholder]
+     (cond
+       (< max-depth depth) dots
+       (seq? v) (let [[pre suf] (split-at max-count v)]
+                  (doall
+                    (concat (map tru pre)
+                            (when (seq suf)
+                              [dots]))))
+       (set? v) (let [pre (take max-count v)
+                      ;; in case keys conflict after truncation,
+                      ;; remove all conflicting keys.
+                      ;; eg., if kvs is [:a :a :b]
+                      ;;      then truncate to [:b]
+                      ks (mapv tru pre)
+                      s (->> ks
+                             frequencies
+                             (filter (comp #{1} second))
+                             (map first)
+                             set)]
+                  (cond-> s
+                    (not= (count s) (count v)) (conj dots)))
```

Comment:
I would truncate the values first, remove conflicting ones and then remove elements.
``` clojure
(let [s (->> (map tru pre)
             set
             (take max-count))]
  (cond-> s
      (not= (count s) (count v)) (conj (vec dots))))
```

Reply:
BTW, here we change the type of the value, it's not a big deal when we just look at it, but it could impact an agent that would reason on these events.

Reply:
Hmm I think you're confusing this case with the `vector?` case. That's at the bottom.

Reply:
Sorry, I think that this is equivalent to your code
``` clojure
(let [s (->> (take max-count)
                   (map tru pre)
                   set)]
  (cond-> s
      (not= (count s) (count v)) (conj dots)))
```

In any case, the result could be weird for a set since it's unordered:
``` clojure
user=> (conj #{1 2 3} 4)
#{1 4 3 2}
user=> (conj #{1 2 3} "...")
#{"..." 1 3 2}
user=> (conj #{"el1" "el2" "el3"} "...")
#{"..." "el3" "el1" "el2"}
```